### PR TITLE
feat: add the confirmed events in the highest layers

### PR DESCRIPTION
**Pull request by vrolland**
## Description of the changes

Propagate the `confirmed` event to the request-logic

Next steps:
- [x] propagate the event to the request-client.js
- [x] add confirmed mechanism to the node
- [x] update integration tests

### CODE REVIEW

File: packages/request-client.js/src/http-data-access.ts [hunk 1]
```diff
@@ -64,7 +66,16 @@ export default class HttpDataAccess implements DataAccessTypes.IDataAccess {
       },
       this.axiosConfig,
     );
-    return data;
+
+    const result: DataAccessTypes.IReturnPersistTransaction = Object.assign(
```
Comment (rittme) [inline packages/request-client.js/src/http-data-access.ts]: comment?

File: packages/request-client.js/src/http-metamask-data-access.ts [hunk 2]
```diff
@@ -136,14 +137,38 @@ export default class HttpMetaMaskDataAccess extends HttpDataAccess {
...
       meta: {
         storageMeta,
         topics: topics || [],
         transactionStorageLocation: ipfsHash,
       },
       result: {},
-    };
+    });
+
+    tx.wait().then((txConfirmed: any) => {
```
Comment (rittme) [inline packages/request-client.js/src/http-metamask-data-access.ts]: A comment is quite important here.

File: packages/request-client.js/src/http-metamask-data-access.ts [hunk 3]
```diff
@@ -136,14 +137,38 @@ export default class HttpMetaMaskDataAccess extends HttpDataAccess {
...
+        blockConfirmation: txConfirmed.confirmations,
+        blockNumber: txConfirmed.blockNumber,
+        blockTimestamp: ethBlock.timestamp,
+        fee,
+        networkName: this.networkName,
+        smartContractAddress: txConfirmed.to,
+        transactionHash: txConfirmed.hash,
+      };
+
+      result.emit('confirmed', {
```
Comment (rittme) [inline packages/request-client.js/src/http-metamask-data-access.ts]: comment here too?

File: packages/request-logic/src/request-logic.ts [hunk 4]
```diff
@@ -56,10 +58,23 @@ export default class RequestLogic implements RequestLogicTypes.IRequestLogic {
...
     );
-    return {
+
+    const result = Object.assign(new EventEmitter(), {
       meta: { transactionManagerMeta: resultPersistTx.meta },
       result: { requestId },
-    };
+    });
+
+    resultPersistTx.on(
```
Comment (rittme) [inline packages/request-logic/src/request-logic.ts]: comment?

File: packages/request-logic/src/request-logic.ts [hunk 5]
```diff
@@ -99,10 +114,23 @@ export default class RequestLogic implements RequestLogicTypes.IRequestLogic {
...
```
Comment (rittme) [inline packages/request-logic/src/request-logic.ts]: comment?

File: packages/request-logic/src/request-logic.ts [hunk 6]
```diff
@@ -165,9 +193,20 @@ export default class RequestLogic implements RequestLogicTypes.IRequestLogic {
       requestId,
     );
 
-    return {
+    const result = Object.assign(new EventEmitter(), {
       meta: { transactionManagerMeta: resultPersistTx.meta },
-    };
+    });
+
+    resultPersistTx.on(
```
Comment (rittme) [inline packages/request-logic/src/request-logic.ts]: comment?

File: packages/request-logic/src/request-logic.ts [hunk 7]
```diff
@@ -201,9 +240,21 @@ export default class RequestLogic implements RequestLogicTypes.IRequestLogic {
...
       requestId,
     );
-    return {
+
+    const result = Object.assign(new EventEmitter(), {
       meta: { transactionManagerMeta: resultPersistTx.meta },
-    };
+    });
+
+    resultPersistTx.on(
```
Comment (rittme) [inline packages/request-logic/src/request-logic.ts]: comment?

File: packages/request-logic/src/request-logic.ts [hunk 8]
```diff
@@ -237,9 +288,21 @@ export default class RequestLogic implements RequestLogicTypes.IRequestLogic {
...
```
Comment (rittme) [inline packages/request-logic/src/request-logic.ts]: comment?

File: packages/request-logic/src/request-logic.ts [hunk 9]
```diff
@@ -273,9 +336,21 @@ export default class RequestLogic implements RequestLogicTypes.IRequestLogic {
...
```
Comment (rittme) [inline packages/request-logic/src/request-logic.ts]: comment?

File: packages/request-logic/src/request-logic.ts [hunk 10]
```diff
@@ -310,9 +385,21 @@ export default class RequestLogic implements RequestLogicTypes.IRequestLogic {
...
```
Comment (rittme) [inline packages/request-logic/src/request-logic.ts]: comment?

File: packages/transaction-manager/src/transaction-manager.ts [hunk 11]
```diff
@@ -111,13 +113,30 @@ export default class TransactionManager implements TransactionTypes.ITransaction
...
+    const result: TransactionTypes.IReturnPersistTransaction = Object.assign(new EventEmitter(), {
       meta: {
         dataAccessMeta: persistResult.meta,
         encryptionMethod: channelEncryptionMethod,
       },
       result: {},
-    };
+    });
+
+    persistResult.on(
```
Comment (rittme) [inline packages/transaction-manager/src/transaction-manager.ts]: comment?

File: packages/data-access/test/data-access.test.ts [hunk 12]
```diff
@@ -415,17 +450,19 @@ describe('data-access', () => {
...
-          storageMeta: {
-            state: DataAccessTypes.TransactionState.PENDING,
-            timestamp: 1,
-          },
-          topics: [arbitraryTopic1],
-          transactionStorageLocation: dataIdBlock2tx,
-        },
-        result: {},
-      });
+      // expect(result, 'result wrong').to.deep.equal(
```
Comment (ltacker) [inline packages/data-access/test/data-access.test.ts]: Use `.skip`

Comment (ltacker) [inline packages/data-access/test/data-access.test.ts]: Why is it put into comment ?

File: packages/request-node/src/request/confirmedTransactionStore.ts [hunk 13]
```diff
@@ -0,0 +1,66 @@
+import { DataAccessTypes, LogTypes } from '@requestnetwork/types';
+import * as httpStatus from 'http-status-codes';
+
+import * as Keyv from 'keyv';
+
+/**
+ * Class for storing confirmed transaction information
```
Comment (rittme) [inline packages/request-node/src/request/confirmedTransactionStore.ts]:
```suggestion
 * Class for storing confirmed transactions information
```

File: packages/request-node/src/request/confirmedTransactionStore.ts [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+ * This class allow to store and serve the information when a transaction is confirmed
```
Comment (rittme) [inline packages/request-node/src/request/confirmedTransactionStore.ts]: This is not necessary, it's a duplicate from the previous line.

File: packages/request-node/src/request/confirmedTransactionStore.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+ */
+export default class ConfirmedTransactionStore {
+  public store: Keyv<DataAccessTypes.IReturnPersistTransaction>;
+
+  /**
+   * Confirmed transaction store constructor
```
Comment (rittme) [inline packages/request-node/src/request/confirmedTransactionStore.ts]:
```suggestion
   * Confirmed transactions store constructor
```

File: packages/request-node/src/request/confirmedTransactionStore.ts [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+   */
+  constructor() {
+    this.store = new Keyv<DataAccessTypes.IReturnPersistTransaction>();
```
Comment (rittme) [inline packages/request-node/src/request/confirmedTransactionStore.ts]: The store seems to be in-memory only. Is this what we want? (I'm not suggesting it's not, just want to make sure)

File: packages/request-node/src/request/confirmedTransactionStore.ts [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+   * @param serverResponse http server response object
+   * @param logger logger
+   */
+  public async getConfirmedTransaction(
+    clientRequest: any,
+    serverResponse: any,
+    logger: LogTypes.ILogger,
+  ): Promise<void> {
+    if (!clientRequest.query.transactionHash) {
+      serverResponse.status(httpStatus.UNPROCESSABLE_ENTITY).send('Incorrect data');
```
Comment (rittme) [inline packages/request-node/src/request/confirmedTransactionStore.ts]: The error message could be clearer.

File: packages/request-node/src/request/persistTransaction.ts [hunk 18]
```diff
@@ -1,56 +1,92 @@
...
 
+import ConfirmedTransactionStore from './confirmedTransactionStore';
+
 /**
- * Handles persistTransaction of data-access layer.
- *
- * @param clientRequest http client request object
- * @param serverResponse http server response object
- * @param dataAccess data access layer
+ * Class to persist transaction though data-access layer
```
Comment (rittme) [inline packages/request-node/src/request/persistTransaction.ts]:
```suggestion
 * Class to persist transactions though the data-access layer
```

File: packages/request-client.js/src/http-data-access.ts [hunk 19]
```diff
@@ -2,12 +2,20 @@ import { DataAccessTypes } from '@requestnetwork/types';
...
 
+import { EventEmitter } from 'events';
+
 // Maximum number of retries to attempt when http requests to the Node fail
 const HTTP_REQUEST_MAX_RETRY = 3;
 
 // Delay between retry in ms
 const HTTP_REQUEST_RETRY_DELAY = 100;
 
+// Maximum number of retries to attempt when http requests to the Node fail
```
Comment (ltacker) [inline packages/request-client.js/src/http-data-access.ts]: I don't really what does it mean "getting the confirmation" compared to the http request, maybe this comment need more clarity

File: packages/request-node/src/request/confirmedTransactionStore.ts (hunk 14 above)
Comment (ltacker) [reply packages/request-node/src/request/confirmedTransactionStore.ts]: I think we need more clarity, why do we need a class to store confirmed transaction ?
What `serve the information` does mean ?

File: packages/types/src/request-logic-types.ts [hunk 20]
```diff
@@ -95,7 +95,6 @@ export interface IReturnCreateRequest extends IRequestLogicReturnWithConfirmatio
 
 /** return of the function getFirstRequestFromTopic */
 export interface IReturnGetRequestFromId extends IRequestLogicReturn {
-  // TODO replace any by a better type
```
[the remark this answers is not in the document]
Comment (vrolland) [reply packages/types/src/request-logic-types.ts]: you are right. I've updated it.

File: packages/request-node/src/request/confirmedTransactionStore.ts (hunk 14 above)
Comment (vrolland) [reply packages/request-node/src/request/confirmedTransactionStore.ts]: I rephrased it

File: packages/request-node/src/request/confirmedTransactionStore.ts (hunk 16 above)
Comment (vrolland) [reply packages/request-node/src/request/confirmedTransactionStore.ts]: As this is just a temporary storing, (after the client succeed to call this data they become useless), in-memory is enough. 
We will need a garbage collecter though, to not keep for ever useless data. (I jira It)

File: packages/request-client.js/test/index.test.ts [hunk 21]
```diff
@@ -281,16 +289,19 @@ describe('index', () => {
...
       signatureProvider: fakeSignatureProvider,
     });
-    await requestNetwork.createRequest({
+    const request = await requestNetwork.createRequest({
       requestInfo: TestData.parametersWithoutExtensionsData,
       signer: payeeIdentity,
     });
     expect(spy).to.have.been.called.once;
+
+    await new Promise((resolve): any => request.on('confirmed', resolve));
```
Comment (benjlevesque) [inline packages/request-client.js/test/index.test.ts]: this is repeated a lot, maybe add a new method?
```typescript
public waitForConfirmation(): Promise<IRequestData>{
    return new Promise(resolve => this.on('confirmed', resolve);
}
```
usage: `await request.waitForConfirmation()`

File: packages/request-client.js/test/index.test.ts [hunk 22]
```diff
@@ -413,9 +431,17 @@ describe('index', () => {
 
     const data = request.getData();
     expect(data).to.exist;
-    expect(data.balance).to.be.null;
-    expect(data.meta).to.exist;
+    expect(data.balance).to.null;
+    expect(data.meta).to.be.exist;
```
Comment (benjlevesque) [inline packages/request-client.js/test/index.test.ts]:
```suggestion
    expect(data.meta).to.exist;
```

File: packages/types/src/client-types.ts [hunk 23]
```diff
@@ -16,6 +16,9 @@ export interface IRequestData extends Omit<RequestLogic.IRequest, 'currency'> {
   contentData: any;
   currencyInfo: RequestLogic.ICurrency;
   pending: RequestLogic.IPendingRequest | null;
+
+  on: (type: any, listener: any) => this;
```
[the remark this answers is not in the document]
Comment (benjlevesque) [reply packages/types/src/client-types.ts]: https://stackoverflow.com/a/39145058

File: packages/request-node/src/request/confirmedTransactionStore.ts (hunk 16 above)
Comment (benjlevesque) [reply packages/request-node/src/request/confirmedTransactionStore.ts]: what happens if the node is restarted ? What's the expected behavior of the `/getConfirmedTransaction` endpoint 
- for a non-confirmed tx
- for a confirmed tx

File: packages/request-client.js/src/http-data-access.ts [hunk 24]
```diff
@@ -2,12 +2,20 @@ import { DataAccessTypes } from '@requestnetwork/types';
...
+const GET_CONFIRMATION_MAX_RETRY = 500;
+
+// Delay between retry in ms
+const GET_CONFIRMATION_RETRY_DELAY = 3000;
```
[the remark this answers is not in the document]
Comment (benjlevesque) [reply packages/request-client.js/src/http-data-access.ts]: discussed ☎️

File: packages/request-node/src/request/confirmedTransactionStore.ts (hunk 16 above)
Comment (vrolland) [reply packages/request-node/src/request/confirmedTransactionStore.ts]: discussed. Ok for now. Jira for a long term solution.

**Pull request merged**